File: MainActivity.java
package com.example.weatherapp;

import android.app.AlertDialog;
import android.database.Cursor;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Looper;
import android.util.Log;

import com.example.weatherapp.Adapter.ViewPagerAdapter;
import com.example.weatherapp.Common.Common;
import com.example.weatherapp.Model.WeatherResult;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.tabs.TabLayout;
import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import android.content.Intent;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;
import android.view.LayoutInflater;

import java.util.List;

//import test

import android.os.Bundle;

import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.example.weatherapp.Common.Common;
import com.example.weatherapp.Model.WeatherResult;
import com.example.weatherapp.Retrofit.IOpenWeatherMap;
import com.example.weatherapp.Retrofit.RetrofitClient;
import com.squareup.picasso.Picasso;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;
//test


public class MainActivity extends AppCompatActivity {

    private Toolbar toolbar;
    private TabLayout tabLayout;
    private ViewPager viewPager;

    private CoordinatorLayout coordinatorLayout;

    private FusedLocationProviderClient fusedLocationProviderClient;
    private LocationCallback locationCallback;
    private LocationRequest locationRequest;


    DatabaseHelper myDb;
    Button btnSaveData;
    Button btnViewAll;
    //Button btnDelete;
    //Button btnviewUpdate;

    //test

    class TodayWeatherFragment extends Fragment{

        ImageView img_weather;
        TextView txt_city_name,txt_humidity,txt_sunrise,txt_sunset,txt_pressure,txt_temperature,txt_description,txt_date_time,txt_wind,txt_geo_coord;
        LinearLayout weather_panel;
        ProgressBar loading;


        CompositeDisposable compositeDisposable;
        IOpenWeatherMap mService;

         TodayWeatherFragment instance;

        public  TodayWeatherFragment getInstance() {
            if(instance == null)
                instance = new TodayWeatherFragment();
            return instance;
        }

        public TodayWeatherFragment() {
            compositeDisposable = new CompositeDisposable();
            Retrofit retrofit = RetrofitClient.getInstance();
            mService = retrofit.create(IOpenWeatherMap.class);
        }


        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                 Bundle savedInstanceState) {
            // Inflate the layout for this fragment
            View itemView =  inflater.inflate(R.layout.activity_main, container, false);

            img_weather = (ImageView)itemView.findViewById(R.id.image_weather);
            txt_city_name = (TextView)itemView.findViewById(R.id.txt_city_name);
            txt_humidity = (TextView)itemView.findViewById(R.id.txt_humidity);
            txt_sunrise = (TextView)itemView.findViewById(R.id.txt_sunrise);
            txt_sunset = (TextView)itemView.findViewById(R.id.txt_sunset);
            txt_pressure = (TextView)itemView.findViewById(R.id.txt_pressure);
            txt_temperature = (TextView)itemView.findViewById(R.id.txt_temperature);
            txt_description = (TextView)itemView.findViewById(R.id.txt_description);
            txt_date_time = (TextView)itemView.findViewById(R.id.txt_date_time);
            txt_wind = (TextView)itemView.findViewById(R.id.txt_wind);
            txt_geo_coord = (TextView)itemView.findViewById(R.id.txt_coords);


            getWeatherInformation();

            return itemView;
        }

        private void getWeatherInformation() {
            compositeDisposable.add(mService.getWeatherByLating(String.valueOf(Common.current_location.getLatitude()),
                    String.valueOf(Common.current_location.getLongitude()),
                    Common.APP_ID,
                    "metric")
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new Consumer<WeatherResult>() {
                        @Override
                        public void accept(WeatherResult weatherResult) throws Exception {

                            //Load image
                            Picasso.get().load(new StringBuilder("https://openweathermap.org/img/wn/")
                                    .append(weatherResult.getWeather().get(0).getIcon())
                                    .append(".png").toString()).into(img_weather);


                            //Load info
                            txt_city_name.setText(weatherResult.getName());
                            txt_description.setText(new StringBuilder("Weather in ")
                                    .append(weatherResult.getName()).toString());
                            txt_temperature.setText(new StringBuilder(String.valueOf(weatherResult.getMain().getTemp()))
                                    .append("°C").toString());
                            txt_date_time.setText(Common.convertUnixToDate(weatherResult.getDt()));
                            txt_pressure.setText(new StringBuilder(String.valueOf(weatherResult.getMain().getPressure()))
                                    .append(" hpa").toString());
                            txt_humidity.setText(new StringBuilder(String.valueOf(weatherResult.getMain().getHumidity()))
                                    .append(" %").toString());
                            txt_sunrise.setText(Common.convertUnixToHour(weatherResult.getSys().getSunrise()));
                            txt_sunset.setText(Common.convertUnixToHour(weatherResult.getSys().getSunset()));
                            txt_geo_coord.setText(new StringBuilder(weatherResult.getCoord().toString()).toString());

                            //Display panel
                            weather_panel.setVisibility(View.VISIBLE);
                            loading.setVisibility(View.GONE);



                        }
                    }, new Consumer<Throwable>() {
                        @Override
                        public void accept(Throwable throwable) throws Exception {
                            Toast.makeText(getActivity(), ""+throwable.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    })

            );


        }

    }

    //end of test
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        myDb = new DatabaseHelper(this);

        coordinatorLayout = (CoordinatorLayout) findViewById(R.id.root_view);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        //Request Permission
        Dexter.withActivity(this)
                .withPermissions(Manifest.permission.ACCESS_COARSE_LOCATION,
                        Manifest.permission.ACCESS_FINE_LOCATION)
                .withListener(new MultiplePermissionsListener() {
                    @Override
                    public void onPermissionsChecked(MultiplePermissionsReport report) {
                        if (report.areAllPermissionsGranted()) {
                            buildLocationRequest();
                            buildLocationCallBack();

                            if (ActivityCompat.checkSelfPermission(MainActivity.this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(MainActivity.this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                                return;
                            }
                            fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(MainActivity.this);
                            fusedLocationProviderClient.requestLocationUpdates(locationRequest, locationCallback, Looper.myLooper());
                        }
                    }

                    @Override
                    public void onPermissionRationaleShouldBeShown(List<PermissionRequest> permissions, PermissionToken token) {
                        Snackbar.make(coordinatorLayout, "Permission Denied", Snackbar.LENGTH_LONG)
                                .show();
                    }
                }).check();

        //test from DB project

        btnSaveData = (Button) findViewById(R.id.btnSave);
        btnViewAll = (Button) findViewById(R.id.btnView);

        //btnviewUpdate= (Button)findViewById(R.id.button_update);
        // btnDelete= (Button)findViewById(R.id.button_delete);
        SaveData();
        viewAll();
        //UpdateData();
        //DeleteData();

    }

    private void SaveData() {
        btnSaveData.setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        boolean isInserted = myDb.insertData(//editDate.getText().toString(),
                                txt_temperature.getText());
                                /*editWeather.getText().toString(),
                                editTextcity.getText().toString());*/
                        if (isInserted == true)
                            Toast.makeText(MainActivity.this, "Data Inserted", Toast.LENGTH_LONG).show();
                        else
                            Toast.makeText(MainActivity.this, "Data not Inserted", Toast.LENGTH_LONG).show();
                    }
                }
        );
    }

    public void viewAll() {
        btnViewAll.setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Cursor res = myDb.getAllData();
                        if (res.getCount() == 0) {
                            // show message
                            showMessage("Error", "Nothing found");
                            return;
                        }

                        StringBuffer buffer = new StringBuffer();
                        while (res.moveToNext()) {
                            buffer.append("Date :" + res.getString(0) + "\n");
                            buffer.append("Temperature :" + res.getString(1) + "\n");
                            buffer.append("Weather :" + res.getString(2) + "\n");
                            buffer.append("City :" + res.getString(3) + "\n\n");
                        }

                        // Show all data
                        showMessage("Data", buffer.toString());
                    }
                }
        );
    }

    public void showMessage(String title, String Message) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setCancelable(true);
        builder.setTitle(title);
        builder.setMessage(Message);
        builder.show();
    }

    private void buildLocationCallBack() {
        locationCallback = new LocationCallback() {

            @Override
            public void onLocationResult(LocationResult locationResult) {
                super.onLocationResult(locationResult);

                Common.current_location = locationResult.getLastLocation();

                viewPager = (ViewPager) findViewById(R.id.viewPager);
                setupViewPager(viewPager);
                tabLayout = (TabLayout) findViewById(R.id.tabs);
                tabLayout.setupWithViewPager(viewPager);

                //Log
                Log.d("Location", locationResult.getLastLocation().getLatitude() + "/" + locationResult.getLastLocation().getLongitude());
            }
        };
    }

    private void setupViewPager(ViewPager viewPager) {
        ViewPagerAdapter adapter = new ViewPagerAdapter(getSupportFragmentManager());
        adapter.addFragment(TodayWeatherFragment.getInstance(), "Today");
        adapter.addFragment(ForecastFragment.getInstance(), "5 DAYS");
        viewPager.setAdapter(adapter);
    }

    private void buildLocationRequest() {
        locationRequest = new LocationRequest();
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
        locationRequest.setInterval(5000);
        locationRequest.setFastestInterval(3000);
        locationRequest.setSmallestDisplacement(10.0f);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        switch (item.getItemId()) {
            case R.id.action_search:
                startActivity(new Intent(MainActivity.this, SearchActivity.class));
                return true;
            case R.id.action_location:
                //kwdikas gia na kanei track me to gps thn perioxh toy xrhsth
                return true;
            case R.id.action_settings:
                startActivity(new Intent(MainActivity.this, SettingsActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

    //test

File: MainActivity.kt
package com.cammace.aurora.ui

import android.Manifest
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.cammace.aurora.*
import com.cammace.aurora.databinding.ActivityMainBinding
import com.cammace.aurora.utils.WeatherIcons
import com.cammace.aurora.viewmodel.MainActivityViewModel
import com.mapbox.api.geocoding.v5.GeocodingCriteria
import com.mapbox.mapboxsdk.plugins.places.autocomplete.model.PlaceOptions
import com.mapbox.mapboxsdk.plugins.places.autocomplete.ui.PlaceAutocompleteFragment
import kotlinx.android.synthetic.main.activity_main.*
import timber.log.Timber
import java.text.SimpleDateFormat
import java.util.*



const val REQUEST_COARSE_LOCATION = 5678

class MainActivity : AppCompatActivity() {

  private lateinit var binding: ActivityMainBinding
  private lateinit var viewModel: MainActivityViewModel

  private var weatherIconMap: Map<String, Drawable>? = null

  // Places autocomplete fragment
  private var placeAutoCompleteFragment: PlaceAutocompleteFragment? = null

  // 5 day forecast adapter
  private val adapter = DailyForecastAdapter()

  //for DB
  var myDb: DatabaseHelper? = null
  var btnSaveData: Button? = null
  var btnviewAll: Button? = null
  var Temperature: TextView? = null
  var Weather: TextView? = null


  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
    viewModel = ViewModelProviders.of(this).get(MainActivityViewModel::class.java)
    addObservers()

    // Setup toolbar and hide title
    setSupportActionBar(toolbar_mainActivity)
    supportActionBar?.setDisplayShowTitleEnabled(false)

    weatherIconMap = WeatherIcons.map(this)

    var Latitude: String?
    var Longtitude: String?

    Latitude = intent.getStringExtra("Lat")
    Longtitude = intent.getStringExtra("Long")

    if(Latitude != null && Longtitude != null ){
      viewModel.useCoordinates(Latitude.toDouble(),Longtitude.toDouble())
    }
    else {
      viewModel.getUsersCurrentLocation()
    }

    recyclerView_mainActivity_dailyForecast.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
    recyclerView_mainActivity_dailyForecast.adapter = adapter

    //onCreate from DB

    myDb = DatabaseHelper(this)
    Temperature = findViewById<TextView>(R.id.textView_mainActivity_temp)
    Weather = findViewById<TextView>(R.id.textView_mainActivity_summary)
    btnSaveData = findViewById<View>(R.id.button_save) as Button
    btnviewAll = findViewById<View>(R.id.button_viewAll) as Button
    saveData()
    viewAll()
  }

  private fun saveData() {
    btnSaveData!!.setOnClickListener(
            View.OnClickListener {
              val date = Date()
              val formatter = SimpleDateFormat("dd-MM-yyyy HH:mm:ss")
              val isInserted = myDb!!.insertData(formatter.format(date),Temperature?.text.toString(),Weather?.text.toString(),binding.locationName)
              if (isInserted == true)
                  Toast.makeText(this@MainActivity, "Data Inserted", Toast.LENGTH_LONG).show()
              else
                  Toast.makeText(this@MainActivity, "Data not Inserted", Toast.LENGTH_LONG).show()
            }
    )
  }

  fun viewAll() {
    btnviewAll!!.setOnClickListener(
            View.OnClickListener {
              val res = myDb!!.allData
              if (res.count == 0) { // show message
                showMessage("Error", "Nothing found")
                return@OnClickListener
              }
              val buffer = StringBuffer()
              while (res.moveToNext()) {
                buffer.append("Id : " + res.getString(0) + "\n")
                buffer.append("Date : " + res.getString(1) + "\n")
                buffer.append("Temperature : " + res.getString(2) + "\n")
                buffer.append("Weather : " + res.getString(3) + "\n")
                buffer.append("City : " + res.getString(4) + "\n\n")
              }
              // Show all data
              showMessage("Data", buffer.toString())
            }
    )
  }

  private fun showMessage(title: String?, Message: String?) {
    val builder = AlertDialog.Builder(this)
    builder.setCancelable(true)
    builder.setTitle(title)
    builder.setMessage(Message)
    builder.show()
  }

  override fun onCreateOptionsMenu(menu: Menu?): Boolean {
    menuInflater.inflate(R.menu.menu_toolbar, menu)
    return super.onCreateOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem?): Boolean {
    if (item?.itemId == R.id.action_search) {
      displayPlacesFragment()
    } else if (item?.itemId == R.id.action_user_location) {
      viewModel.getUsersCurrentLocation()
    }
    else if (item?.itemId == R.id.action_about){
      startActivity(Intent(this@MainActivity, AboutActivity::class.java))
    }
    else {
      displayPlacesFragmentTest()
    }
    return super.onOptionsItemSelected(item)
  }

  override fun onBackPressed() {
    if (placeAutoCompleteFragment != null && placeAutoCompleteFragment!!.isVisible) {
      hidePlacesFragment()
      return
    }
    super.onBackPressed()
  }

  private fun displayPlacesFragment() {
    if (placeAutoCompleteFragment == null) {

      val options = PlaceOptions.builder()
        .toolbarColor(ContextCompat.getColor(this, R.color.materialGray_50))
        .backgroundColor(ContextCompat.getColor(this, R.color.materialGray_50))
        .geocodingTypes(GeocodingCriteria.TYPE_PLACE, GeocodingCriteria.TYPE_REGION)
        .build()

      window.statusBarColor = ContextCompat.getColor(this, R.color.materialGray_500)
      placeAutoCompleteFragment = PlaceAutocompleteFragment.newInstance(BuildConfig.MAPBOX_ACCESS_TOKEN, options)
      placeAutoCompleteFragment!!.setOnPlaceSelectedListener(viewModel)

      val transaction = supportFragmentManager.beginTransaction()
      transaction.add(R.id.fragment_container, placeAutoCompleteFragment!!, PlaceAutocompleteFragment.TAG)
      transaction.commit()
    } else {

      showPlacesFragment()
    }
  }

  private fun showPlacesFragment() {
    window.statusBarColor = ContextCompat.getColor(this, R.color.materialGray_500)
    supportFragmentManager.beginTransaction().show(placeAutoCompleteFragment!!).commit()
  }

  //test
  private fun displayPlacesFragmentTest() {
    val intent = Intent(this, coordinates::class.java)
    startActivity(intent)
  }

  private fun hidePlacesFragment() {
    window.statusBarColor = ContextCompat.getColor(this, R.color.colorPrimaryDark)
    val inputMethod = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
    inputMethod.hideSoftInputFromWindow(textView_mainActivity_wind.windowToken, 0)
    supportFragmentManager.beginTransaction().hide(placeAutoCompleteFragment!!).commit()
  }

  private fun addObservers() {
    viewModel.requestLocationPermissionLiveData.observe(this, Observer { shouldRequestPermission ->
      if (shouldRequestPermission) {
        Timber.v("requesting permission")
        requestPermissions()
      }
    })

    viewModel.locationNameLiveData.observe(this, Observer { locationName ->
      binding.locationName = locationName
    })

    viewModel.darkSkyApiResponseLiveData.observe(this, Observer { darkSkyModel ->
      binding.currentCondition = darkSkyModel.currently

      adapter.setDayForecast(darkSkyModel.daily.data)

      // Bind the current weather icon
      if (darkSkyModel.currently.icon != null && weatherIconMap != null) {
        binding.currentConditionIcon = weatherIconMap!![darkSkyModel.currently.icon]
      }
    })

    viewModel.userFinishedSearchLiveData.observe(this, Observer { canceled ->
      Timber.v("User has canceled geocoding search.")
      if (canceled && placeAutoCompleteFragment != null) {
        hidePlacesFragment()
      }
    })
  }

  private fun requestPermissions() {
    ActivityCompat.requestPermissions(this,
      arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION), REQUEST_COARSE_LOCATION)
  }

  override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
    if (requestCode == REQUEST_COARSE_LOCATION && grantResults.isNotEmpty()
      && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
      Timber.v("User gave location permission, continue with getting user's last location.")
      viewModel.getUsersCurrentLocation()
    } else {
      Timber.v("User refused to give location permission. Continue using the default location.")
    }
  }
}
